### PR TITLE
More precise tx output range error reporting

### PR DESCRIPTION
**Pull request by dcoutts**
Report negative tx output quantities separately from the case of too
large, and report the specific quantity (useful for the multi-asset case
where there may be many quantities).

### CODE REVIEW

**Comment by dcoutts** [comment]
Poor error message as spotted by @mkoura

**Review by newhoggy — APPROVED** [review]

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 1]
```diff
@@ -1058,9 +1083,16 @@ makeShelleyTransactionBody era@ShelleyBasedEraMary
...
+      [ do allPositive
+           allWithinMaxBound
+      | let ub = fromIntegral (maxBound :: Word64) :: Quantity
+      , txout@(TxOut _ (TxOutValue MultiAssetInMaryEra v)) <- txOuts
+      , let allPositive       = case [ q | (_,q) <- valueToList v, q >= 0 ] of
+                                  []  -> Right ()
+                                  q:_ -> Left (TxBodyOutputNegative q txout)
+            allWithinMaxBound = case [ q | (_,q) <- valueToList v, q < ub ] of
+                                  []  -> Right ()
+                                  q:_ -> Left (TxBodyOutputOverflow q txout)
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/TxBody.hs:1095]: I think these conditions are backwards.  Should be `q < 0` and `q >= ub`.

**Comment by newhoggy** [comment]
Also I have a question about `maxBound :: Word64`.  Is this a valid txout value?  If so, we should probably use `v <= ub` instead of `v < ub`.  Disregard if `maxBound` is some sort of sentinel value and not a value txout value.

**Review by newhoggy — APPROVED** [review]

**Review by intricate — APPROVED** [review]

**Comment by newhoggy** [comment]
bors merge

**Pull request merged**